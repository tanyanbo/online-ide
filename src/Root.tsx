import React from "react";
import { Provider } from "react-redux";
import { legacy_createStore } from "redux";
import reducers from "./reducers";

let store = legacy_createStore(reducers);

const Root = ({ children, initialState = {} }) => {
  store = legacy_createStore(reducers, initialState);
  return <Provider store={store}>{children}</Provider>;
};

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default Root;
