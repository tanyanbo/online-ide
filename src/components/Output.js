import React, { useEffect, useRef } from "react";
import { connect } from "react-redux";
import { changeRunWhileTyping, changeIsRunning } from "../actions";
import OUTPUT_STRING from "../shared/outputString";

const Output = (props) => {
  const iframeRef = useRef(null);

  useEffect(() => {
    iframeRef.current.contentWindow.onerror = function () {
      console.log("my own error from iframe srcdoc");
    };
  }, []);

  const { changeIsRunning, changeRunWhileTyping } = props;
  const srcDoc = OUTPUT_STRING[props.languages](
    props.html,
    props.css,
    props.js
  );
  const changeRunState = (e) => {
    changeRunWhileTyping(e.target.checked);
  };

  const run = () => {
    changeIsRunning(true);
  };

  return (
    <>
      <div className="buttons">
        <form className="run-checkbox">
          <input type="checkbox" id="run" onChange={changeRunState} />
          <label htmlFor="run">边写边运行</label>
        </form>
        <button className="run-button" disabled={props.checked} onClick={run}>
          运行
        </button>
      </div>
      <iframe
        ref={iframeRef}
        srcDoc={srcDoc}
        sandbox="allow-scripts allow-same-origin"
        title="output"
        height="500px"
        width="100%"
      />
    </>
  );
};

const mapStateToProps = (state) => {
  return {
    html: state.srcDoc.html,
    css: state.srcDoc.css,
    js: state.srcDoc.js,
    checked: state.checkbox.checked,
    languages: state.languages.languages,
  };
};

export default connect(mapStateToProps, {
  changeRunWhileTyping,
  changeIsRunning,
})(Output);
