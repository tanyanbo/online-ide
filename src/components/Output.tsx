import React, { useEffect, useRef } from "react";
import { connect, ConnectedProps } from "react-redux";
import { changeRunWhileTyping, changeIsRunning } from "../actions";
import OUTPUT_STRING from "../shared/outputString";
import { RootState } from "../Root";

const mapStateToProps = (state: RootState) => {
  return {
    html: state.srcDoc.html,
    css: state.srcDoc.css,
    js: state.srcDoc.js,
    checked: state.checkbox.checked,
    languages: state.languages.languages,
  };
};

const connector = connect(mapStateToProps, {
  changeRunWhileTyping,
  changeIsRunning,
});

type Props = ConnectedProps<typeof connector>;

const Output = (props: Props) => {
  const iframeRef = useRef(null);

  useEffect(() => {
    console.log("changed");
    console.log(iframeRef.current.contentWindow);
    iframeRef.current.contentWindow.onerror = function () {
      console.log("my own error from iframe srcdoc");
    };
  }, []);

  const { changeIsRunning, changeRunWhileTyping } = props;
  const srcDoc = OUTPUT_STRING[props.languages](
    props.html,
    props.css,
    props.js
  );
  const changeRunState = (e: React.ChangeEvent<HTMLInputElement>) => {
    changeRunWhileTyping(e.target.checked);
  };

  const run = () => {
    changeIsRunning(true);
  };

  return (
    <>
      <div className="buttons">
        <form className="run-checkbox">
          <input type="checkbox" id="run" onChange={changeRunState} />
          <label htmlFor="run">边写边运行</label>
        </form>
        <button className="run-button" disabled={props.checked} onClick={run}>
          运行
        </button>
      </div>
      <iframe
        ref={iframeRef}
        srcDoc={srcDoc}
        sandbox="allow-scripts allow-same-origin allow-modals"
        title="output"
        height="500px"
        width="100%"
      />
    </>
  );
};

export default connector(Output);
