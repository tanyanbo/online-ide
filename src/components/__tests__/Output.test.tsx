import React from "react";
import renderer from "react-test-renderer";
import Root from "../../Root";
import Output from "../Output";

describe("<Output />", () => {
  it("renders correct html", () => {
    const component = renderer.create(
      <Root initialState={{ srcDoc: { html: "<h1>Hello World!</h1>" } }}>
        <Output />
      </Root>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correct css", () => {
    const component = renderer.create(
      <Root
        initialState={{
          srcDoc: {
            css: `.first {
                    color: green;
                  }`,
          },
        }}
      >
        <Output />
      </Root>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correct js", () => {
    const component = renderer.create(
      <Root initialState={{ srcDoc: { js: "console.log(10);" } }}>
        <Output />
      </Root>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders everything correctly with language: js", () => {
    const component = renderer.create(
      <Root
        initialState={{
          srcDoc: {
            html: '<h1 class="first">Hello world</h1>',
            css: `.first {
                  color: green;
                  }`,
            js: 'document.querySelector(".first").innerText = 10;',
          },
        }}
      >
        <Output />
      </Root>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders everything correctly with language: ts", () => {
    const component = renderer.create(
      <Root
        initialState={{
          srcDoc: {
            html: '<h1 class="first">Hello world</h1>',
            css: `.first {
                  color: green;
                  }`,
            js: `const a: number = 10;
            document.querySelector(".first").innerText = a;`,
          },
          languages: {
            languages: "HTML+CSS+TS",
          },
        }}
      >
        <Output />
      </Root>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
