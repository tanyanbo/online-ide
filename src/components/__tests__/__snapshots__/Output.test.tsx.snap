// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Output /> renders correct css 1`] = `
Array [
  <div
    className="buttons"
  >
    <form
      className="run-checkbox"
    >
      <input
        id="run"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="run"
      >
        边写边运行
      </label>
    </form>
    <button
      className="run-button"
      disabled={false}
      onClick={[Function]}
    >
      运行
    </button>
  </div>,
  <iframe
    height="500px"
    sandbox="allow-scripts allow-same-origin allow-modals"
    srcDoc="
          <html lang=\\"en-US\\">
          <head>
            <title>Iframe</title>
            <style>
              .first {
                    color: green;
                  }
            </style>
          </head>
          <body>
            undefined
            <script>
              try {
                undefined
              } catch(e) {
                throw new Error('custom error');
              }
            </script>
          </body>
        </html>
      "
    title="output-iframe"
    width="100%"
  />,
]
`;

exports[`<Output /> renders correct html 1`] = `
Array [
  <div
    className="buttons"
  >
    <form
      className="run-checkbox"
    >
      <input
        id="run"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="run"
      >
        边写边运行
      </label>
    </form>
    <button
      className="run-button"
      disabled={false}
      onClick={[Function]}
    >
      运行
    </button>
  </div>,
  <iframe
    height="500px"
    sandbox="allow-scripts allow-same-origin allow-modals"
    srcDoc="
          <html lang=\\"en-US\\">
          <head>
            <title>Iframe</title>
            <style>
              undefined
            </style>
          </head>
          <body>
            <h1>Hello World!</h1>
            <script>
              try {
                undefined
              } catch(e) {
                throw new Error('custom error');
              }
            </script>
          </body>
        </html>
      "
    title="output-iframe"
    width="100%"
  />,
]
`;

exports[`<Output /> renders correct js 1`] = `
Array [
  <div
    className="buttons"
  >
    <form
      className="run-checkbox"
    >
      <input
        id="run"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="run"
      >
        边写边运行
      </label>
    </form>
    <button
      className="run-button"
      disabled={false}
      onClick={[Function]}
    >
      运行
    </button>
  </div>,
  <iframe
    height="500px"
    sandbox="allow-scripts allow-same-origin allow-modals"
    srcDoc="
          <html lang=\\"en-US\\">
          <head>
            <title>Iframe</title>
            <style>
              undefined
            </style>
          </head>
          <body>
            undefined
            <script>
              try {
                console.log(10);
              } catch(e) {
                throw new Error('custom error');
              }
            </script>
          </body>
        </html>
      "
    title="output-iframe"
    width="100%"
  />,
]
`;

exports[`<Output /> renders everything correctly with language: js 1`] = `
Array [
  <div
    className="buttons"
  >
    <form
      className="run-checkbox"
    >
      <input
        id="run"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="run"
      >
        边写边运行
      </label>
    </form>
    <button
      className="run-button"
      disabled={false}
      onClick={[Function]}
    >
      运行
    </button>
  </div>,
  <iframe
    height="500px"
    sandbox="allow-scripts allow-same-origin allow-modals"
    srcDoc="
          <html lang=\\"en-US\\">
          <head>
            <title>Iframe</title>
            <style>
              .first {
                  color: green;
                  }
            </style>
          </head>
          <body>
            <h1 class=\\"first\\">Hello world</h1>
            <script>
              try {
                document.querySelector(\\".first\\").innerText = 10;
              } catch(e) {
                throw new Error('custom error');
              }
            </script>
          </body>
        </html>
      "
    title="output-iframe"
    width="100%"
  />,
]
`;

exports[`<Output /> renders everything correctly with language: ts 1`] = `
Array [
  <div
    className="buttons"
  >
    <form
      className="run-checkbox"
    >
      <input
        id="run"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="run"
      >
        边写边运行
      </label>
    </form>
    <button
      className="run-button"
      disabled={false}
      onClick={[Function]}
    >
      运行
    </button>
  </div>,
  <iframe
    height="500px"
    sandbox="allow-scripts allow-same-origin allow-modals"
    srcDoc="        
        <html lang=\\"en-US\\">
          <head>
            <title>Iframe</title>
            <style>
              .first {
                  color: green;
                  }
            </style>
          </head>
          <body>
            <h1 class=\\"first\\">Hello world</h1>
            <script src=\\"https://unpkg.com/typescript@latest/lib/typescriptServices.js\\"></script>
            <script>
              eval(window.ts.transpile(\`const a: number = 10;
            document.querySelector(\\".first\\").innerText = a;\`));
            </script>
          </body>
        </html>
      "
    title="output-iframe"
    width="100%"
  />,
]
`;
